### DISCUSSION TITLE
Connecting tables cells with arrows

### DISCUSSION THREAD

**Discussion by psads-git** [typst/typst#4510] [Q&A] [answered]
Can the following diagram be done in `typst`, by specifying what are the cells to be connected by arrows?

![image](https://github.com/typst/typst/assets/75945439/c2389a23-08ec-4ad1-8cd6-ebffa76dc20c)

**Comment by MDLC01** [comment]
This looks like a use case for [Diagraph](https://typst.app/universe/package/diagraph/), although you won't have as many table customization features as with native Typst.

**Reply by psads-git** [reply]
Thanks, @MDLC01. for your suggestion. Unfortunately, I need table customization features that  `Diagraph` lacks.

**Comment by TS60** [comment, marked as answer]
Here is one example code that probably can work.

First, clone the table layout:

```typst
#let frame(stroke) = (x, y) => (
  left: if x > 0 { 0pt } else { stroke },
  right: stroke,
  top: if y < 2 { stroke } else {assert(true); 0pt },
  bottom: stroke,
)

#set table(
  //fill: (rgb("EAF2F5"), none),
  stroke: frame(rgb("21222C")),
)



#grid(
  columns: (1fr, 1fr, 1fr),
  gutter: 3pt,
  table(
    columns: 1,
    table.header(
      [XXXXX],
    ),
    [xxxxx1], [xxxxx2], [xxxxx3], [xxxxx4],
  ),
  table(
    columns: 1,
    table.header(
      [YYYYY],
    ),
    [yyyyy1], [yyyyy2], [yyyyy3], [yyyyy4],
  ),
  table(
    columns: 1,
    table.header(
      [ZZZZZ],
    ),
    [zzzzz1], [zzzzz2], [zzzzz3], [zzzzz4], [zzzzz5]
  ),
)
```

![tablelayout](https://github.com/typst/typst/assets/152509161/58d9b3f3-2c5a-40bc-b523-f40513fa27aa)

Next step: Either in the stroke function where you have the (x, y) pairs inject an anchor (see for example the `assert(true)` code that is injected). Or: put the anchor before/behind the cell value like `[zzzzz2]`.

After this, draw an arrow from one anchor (named node) to an other using [cetz](https://raw.githubusercontent.com/cetz-package/cetz/stable/manual.pdf). I don't know enough about this package, but it should work like this.
Maybe someone can make this example complete.

**Comment by psads-git** [comment]
Thanks, @TS60: that is a very interesting idea. Maybe with [`pinit`](https://github.com/OrangeX4/typst-pinit), we could manage to place the `#pin` at the right place in the borders of the tables.

**Comment by psads-git** [comment]
Following your idea, I did following. Only pinning the borders of the tables is missing:

```
#import "@preview/pinit:0.1.4": *

#let frame(stroke) = (x, y) => (
  left: if x > 0 { 0pt } else { stroke },
  right: stroke,
  top: if y < 2 { stroke } else {assert(true); 0pt },
  bottom: stroke,
)

#set table(
  //fill: (rgb("EAF2F5"), none),
  stroke: frame(rgb("21222C")),
)

#grid(
  columns: (1fr, 1fr, 1fr),
  gutter: 3pt,
  table(
    columns: 1,
    table.header(
      [XXXXX],
    ),
    [xxxxx1], [xxxxx2#pin(1)], [xxxxx3], [xxxxx4],
  ),
  table(
    columns: 1,
    table.header(
      [YYYYY],
    ),
    [yyyyy1], [yyyyy2], [#pin(2)yyyyy3], [yyyyy4],
  ),
  table(
    columns: 1,
    table.header(
      [ZZZZZ],
    ),
    [zzzzz1], [zzzzz2], [zzzzz3], [zzzzz4], [zzzzz5]
  ),
)

#pinit-arrow(1, 2)
```

![image](https://github.com/typst/typst/assets/75945439/c0471dc9-67b5-4df1-bbd3-b7bc36a02183)

**Reply by TS60** [reply]
You can fix the positions by hand by using `#pinit-arrow(start-dx: 0pt, start-dy: 0pt, end-dx: 0pt, end-dy: 0pt, 1, 2)` and use appropriate values like -5pt. Maybe you could also find a better text size depending value like -1.5em or so.
Or maybe put something like `#pinit-place` in the stroke function.

**Reply by psads-git** [reply]
That is a good idea, @TS60! Thanks!

Unfortunately, it appears that one cannot draw bidirectional arrows (like <----->). Or am I missing something?

**Reply by TS60** [reply]
Maybe it is possible by setting `tail` of the arrow, but probably very difficult. You can draw the arrow two times (in both directions), but this is visible in the pdf when zooming.

An other idea is to draw a second arrow with zero length (only the arrow head) from and to the same pin. So you have the arrows in both directions, one arrow has the normal line length and the other has the normal arrows head and zero line length. Both could form together the bidirected arrow which is still zoomable. But untested.

**Reply by psads-git** [reply]
Thanks again, @TS60, for your precious help!

I have meanwhile managed to get the double-arrowhead arrow, by creating a hacked version of `pinit-arrow`, which I call `pinit-arrow2`. 

Basically, this `pinit-arrow2` uses `double-arrow` instead of the default `simple-arrow`. Function `double-arrow` adds the second arrowhead.

The code is below and the output too.

```
#import "@preview/pinit:0.1.4": *

#let frame(stroke) = (x, y) => (
  left: if x > 0 { 0pt } else { stroke },
  right: stroke,
  top: if y < 2 { stroke } else {assert(true); 0pt },
  bottom: stroke,
)

#set table(
  //fill: (rgb("EAF2F5"), none),
  stroke: frame(rgb("21222C")),
)

#grid(
  columns: (3cm, 3cm, 3cm),
  gutter: 3pt,
  table(
    columns: 1,
    inset: 3pt, 
    table.header(
      [*XXXXX*],
    ),
    [xxxxx1], [xxxxx2#pin(1)], [xxxxx3], [xxxxx4],
  ),
  table(
    columns: 1,
    inset: 3pt, 
    table.header(
      [*YYYYY*],
    ),
    [yyyyy1], [yyyyy2], [#pin(2)yyyyy3], [yyyyy4],
  ),
  table(
    columns: 1,
    inset: 3pt, 
    table.header(
      [*ZZZZZ*],
    ),
    [zzzzz1], [zzzzz2], [zzzzz3], [zzzzz4], [zzzzz5]
  ),
)

#let double-arrow(
  fill: black,
  stroke: 0em,
  start: (0em, 0em),
  end: (3em, 0em),
  thickness: 0.12em,
  arrow-width: 4,
  arrow-height: 4,
  inset: 0.5,
  tail: (),
) = {
  let _arrow-width = arrow-width * thickness
  let _arrow-height = arrow-height * thickness
  let _inset = inset * thickness
  let start-x = start.at(0)
  let start-y = start.at(1)
  let end-x = end.at(0)
  let end-y = end.at(1)
  let _dx = end-x - start-x
  if _dx.em != 0 and (_dx - _dx.em * 1em).pt() != 0 {
    panic("The x-coordinate must only use pt units or only em units.")
  }
  let _dy = end-y - start-y
  if _dy.em != 0 and (_dy - _dy.em * 1em).pt() != 0 {
    panic("The y-coordinate must only use pt units or only em units.")
  }
  let _angle = if _dx.em != 0 or _dy.em != 0 {
    calc.atan2(_dx.em, _dy.em)
  } else {
    calc.atan2(_dx.pt(), _dy.pt())
  }
  let _ht = 0.5 * thickness
  let _hw = 0.5 * _arrow-width
  let _c = calc.cos(_angle)
  let _s = calc.sin(_angle)
  
  polygon(
    fill: fill,
    stroke: stroke,
    ..tail,
    // First arrowhead points
    (start-x + _s * _ht + _c * (_arrow-height - _inset), start-y - _c * _ht + _s * (_arrow-height - _inset)),
    (start-x + _s * _hw + _c * _arrow-height, start-y - _c * _hw + _s * _arrow-height),
    (start-x, start-y),
    (start-x - _s * _hw + _c * _arrow-height, start-y + _c * _hw + _s * _arrow-height),
    (start-x - _s * _ht + _c * (_arrow-height - _inset), start-y + _c * _ht + _s * (_arrow-height - _inset)),
    // Shaft points
    (start-x + _dx - _s * _ht - _c * (_arrow-height - _inset), start-y + _dy + _c * _ht - _s * (_arrow-height - _inset)),
    // Second arrowhead points
    (start-x + _dx - _s * _hw - _c * _arrow-height, start-y + _dy + _c * _hw - _s * _arrow-height),
    (end-x, end-y),
    (start-x + _dx + _s * _hw - _c * _arrow-height, start-y + _dy - _c * _hw - _s * _arrow-height),
    (start-x + _dx + _s * _ht - _c * (_arrow-height - _inset), start-y + _dy - _c * _ht - _s * (_arrow-height - _inset)),
    // Close the polygon back to the shaft
    (start-x + _s * _ht, start-y - _c * _ht),
  )
}


#let pinit-arrow2(
  start-dx: 0pt,
  start-dy: 0pt,
  end-dx: 0pt,
  end-dy: 0pt,
  start,
  end,
  ..args,
) = {
  pinit((start, end), (locations) => {
    absolute-place(double-arrow(
      start: (
        locations.at(0).x + start-dx,
        locations.at(0).y + start-dy,
      ),
      end: (
        locations.at(1).x + end-dx,
        locations.at(1).y + end-dy,
      ),
      ..args,
    )
    )
  })
}

#pinit-arrow2(1, 2, start-dx: 11pt, start-dy: -2pt, end-dx: -4pt, end-dy: 5pt, thickness: 0.1em)
```

![image](https://github.com/typst/typst/assets/75945439/e5bd271e-3fe7-432f-b6c8-607d31a495e9)

**Reply by TS60** [reply]
This is the best solution!
